docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,2624 @@
+
+[K8S#110714](https://github.com/kubernetes/kubernetes/issues/110714) has been raised.
+
+
+## References
+
+1. https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/
+1. https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/
+
+## Appendix
+
+### Etdc Active-Passive Options
+
+In this topology there will be just one `active/primary etcd` instance and all other `etcd` instances will be running as `hot-standby`. Each `etcd` instance serves as an independent single node cluster.
+There are three options to setup an `active-passive` etcd.
+
+<details>
+<summary>Option-1: Independent single node etcd clusters</summary>
+
+<img src="assets/activepassive-option1.png">
+
+`Primary etcd` will periodically take a snapshot (full and delta) and will push these snapshots to the backup-bucket. `Hot-Standy etcd` instances will periodically query the backup-bucket and sync its database accordingly. If a new full snapshot is available which has a higher revision number than what is available in its local etcd database then it will restore from a full snapshot. It will additionally check if there are delta snapshots having a higher revision number. If that is the case then it will apply the delta snapshots directly to its local etcd database.
+> NOTE: There is no need to run an embedded etcd to apply delta snapshots.
+
+For the sake of illustration only assume that there are two etcd pods `etcd-0` and `etcd-1` with corresponding labels which uniquely identify each pod. Assume that `etcd-0` is the current `primary/active` etcd instance.
+
+`etcd-druid` will take an additional responsibility to monitor the health of `etcd-0` and `etcd-1`. When it detects that the `etcd-0` is no longer healthy it will patch the `etcd` service to point to the `etcd-1` pod by updating the label/selector so that it becomes the `primary` etcd. It will then restart `etc-0` pod and henceforth that will serve as a `hot-standby`.
+
+**Pros**
+* There is no leader election, no quorum related issues to be handled. It is simpler to setup and manage.
+* Allows you to just have a total of two etcd nodes - one is active and another is passive. In an `active-active` cluster it is always recommended to have odd numbered etcd members (minimum of 3). This allows high availability across zones in cases where regions only have 2 zones (e.g. CCloud and Azure regions that do not have more than 2 zones).
+
+**Cons**
+* For a multi-zonal etcd setup since there is only one active etcd, all instances of API Server's spread across zones will connect to a single etcd instance. This results in a drastic increase in cross-zonal traffic, translating to higher operating cost.
```

Comment:
Is that so? In case of our setup, it makes little difference whether the API server or the follower talks to the leader in the other zone – I thought the network measurements showed that?

Reply:
If there are 3 instances of KAPI (one per zone) and each one just talk to one etcd member which is in one zone then there will be an increase in cross zonal traffic from 2 KAPI instances. If there are too many PUT requests to the KAPI in zones other than the etcd then it will surely increase the overall cross-zonal traffic. But you are right - if we compare this with a etcd cluster where each member is in a different zone then for all PUT operations the leader will have to send the payload to all in the etcd cluster anyways adding to comparatively larger cross-zonal network usage. So this con is actually a pro :)
It has been changed.